Minor version update - command restructure

izakp opened the pull request:
Preparing version `0.4.0`

Some old commits in here that made it into `0.3.4` and `0.3.5`

Changes for `0.4.0` release:
- Dependency updates
- Refactoring nesting of commands
- Change `deploy` to `deployment update` and group `history`, `promote` and `refresh` under `deployment` as well
- pre- and post-deploy hooks in project config.yml
- Pod label constraints for `remote run` / ` remote deployment update` / `remote deployment promote` commands
- Waiting for deployment rollout during `deployment update` and `deployment promote`
- Change `dev shell` to `dev run` with command to run as required argument
- Pass through `--tail` option in local and remote logs command in place of `--nlines`
- Add `--internal` option to `hokusai setup` and pass through to `common.build_service` interface

The new structure (correctly using Click's group and command chaining patterns) refactors the command structure into nested subcommands, like so:

**OLD VERSION - SEE UPDATED VERSION BELOW**

```
isacpetruzzi@Artsy-Isac hokusai (master) $ hokusai
Usage: hokusai [OPTIONS] COMMAND [ARGS]...

  Hokusai is a CLI for managing application deployments on Kubernetes

Options:
  --help  Show this message and exit.

Commands:
  check      Check Hokusai dependencies and configuration
  configure  Install and configure kubectl
  local      Interact with your local project and Docker...
  registry   Interact with the project registry
  remote     Interact with remote Kubernetes resources
  setup      Set up Hokusai for the current project
  version    Print Hokusai's version and exit


isacpetruzzi@Artsy-Isac hokusai (master) $ hokusai local
Usage: hokusai local [OPTIONS] COMMAND [ARGS]...

  Interact with your local project and Docker engine

Options:
  --help  Show this message and exit.

Commands:
  build  Build the Docker image defined in...
  dev    Interact with docker-compose targeting the...
  test   Boot the test environment defined in...


isacpetruzzi@Artsy-Isac hokusai (master) $ hokusai local dev
Usage: hokusai local dev [OPTIONS] COMMAND [ARGS]...

  Interact with docker-compose targeting the development environment defined
  in ./hokusai/development.yml

Options:
  --help  Show this message and exit.

Commands:
  clean   Stop and remove all containers in the...
  logs    Print logs from the development environment...
  run     Run a command in the development...
  start   Start the development environment defined in...
  status  Print the status of the development...
  stop    Stop the development environment defined in...


isacpetruzzi@Artsy-Isac hokusai (master) $ hokusai remote
Usage: hokusai remote [OPTIONS] COMMAND [ARGS]...

  Interact with remote Kubernetes resources

Options:
  --help  Show this message and exit.

Commands:
  create      Create the Kubernetes remote resources...
  delete      Delete the Kubernetes remote resources...
  deployment  Interact with the project's' remote...
  env         Interact with the runtime environment for the...
  gitdiff     Print a git diff between the tag currently...
  gitlog      Print a git log between the tag currently...
  logs        Get container logs for a given environment
  run         Launch a new container and run a command in a...
  status      Print the Kubernetes remote resources status...
  update      Update the Kubernetes remote resources...


isacpetruzzi@Artsy-Isac hokusai (master) $ hokusai remote env
Usage: hokusai remote env [OPTIONS] COMMAND [ARGS]...

  Interact with the runtime environment for the application

Options:
  --help  Show this message and exit.

Commands:
  create  Create the Kubernetes configmap object...
  delete  Delete the Kubernetes configmap object...
  get     Print environment variables stored on the...
  set     Set environment variables - each of...
  unset   Unset environment variables - each of...


isacpetruzzi@Artsy-Isac hokusai (master) $ hokusai remote deployment
Usage: hokusai remote deployment [OPTIONS] COMMAND [ARGS]...

  Interact with the project's' remote Kubernetes deployment(s)

Options:
  --help  Show this message and exit.

Commands:
  history  Print the project's deployment history in...
  promote  Update the project's deployment(s) on...
  refresh  Refresh the project's deployment(s) by...
  update   Update the project's deployment(s) for a...
```

This will require updates in CI for downstream projects, but I would like to move them to use CircleCI 2.0 at the same times, as in lewitt-api.

Still need to update docs and unit tests but thought I'd open this as a WIP PR early so y'all can comment on the new nested command structure.

cavvia commented on docs/Getting_Started.md:
````diff
@@ -2,106 +2,116 @@
+```bash
+hokusai local dev status
+hokusai local dev logs
+hokusai local dev run bash
````
one strange thing about the command structure (both before and after this PR), is that environments are top level commands in the `local` context but are flags in the remote context (e.g. `dev` here verses `--production` for remote commands)

cavvia replied:
i don't have a better suggested structure, this is just an observation.

cavvia replied:
if we were to be consistent i guess we would do: 

```
hokusai dev [CMD]
hokusai test
hokusai production [CMD]
hokusai staging [CMD]

cavvia replied:
The issue with my suggested structure is that non environment specific commands (like `promote`) would live at the top-level.

izakp replied:
Yup. This was exactly my line of reasoning, and it's because I didn't want to hoist the promote command to the top level that I didn't try making `staging` and `production` their own command groups. I could give it a go however, and maybe with a bit of metaprogramming get a syntax like `hokusai promote staging to production` WDYT?

cavvia replied:
I think it's pretty intuitive that the `deployment` namespace refers to a remote context actually, I don't have a problem with that living in the top-level. In the structure I'm proposing, you might consider putting `promote` under `staging` as `hokusai staging promote`, but I also think it's fine to group it with the other deployment sub-commands as you do now and have `deployment` live at the top level.

My main issue was with our use of environments as both options and top-level command namespaces and the inconsistency arising from that.

cavvia commented on docs/Getting_Started.md:
```diff
@@ -2,106 +2,116 @@
+  * `./hokusai/development.yml`- a Docker Compose YAML file for the development environment
+  * `./hokusai/test.yml` - a Docker Compose YAML file for the test environment
+  * `./hokusai/staging.yml` - a Kubernetes YAML file for the staging environment
+  * `./hokusai/production.yml` - a Kubernetes YAML file for the production environment
```
You don't need to use the same relative path each time, just file names (`config.yml`) will do.

cavvia commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+
+# Install Hokusai
+ADD hokusai/VERSION.txt VERSION.txt
+RUN pip install hokusai==$(cat VERSION.txt)
```
hokusai can build itself... it's grey goo from here on in.

bhoggard commented:
When @izakp asked me, I said I didn't feel comfortable using names of environments for commands, since people may not use the same set, and we may not eventually. We may have an integration or smoke test environment in addition to staging and production.

cavvia commented:
Good point @bhoggard - I'm still not convinced about `local` and `remote` namespaces either though, as well as our inconsistent use of environments. Happy to go with consensus on this one, so if you're both happy with an alternative structure that's fine by me.

joeyAghion commented:
Speaking as casual `hokusai` _user_ I don't think the `local`/`remote` grouping helps with overall coherence or clarity. And, as you observed already, there are clear exceptions/ambiguities in that scheme. The subcommands where they potentially overlap appear to be `logs`, `run`, and `status`. Some of these already accept arguments defining the target environment, so maybe we should simply support `local` as a special value?

(I tend to agree that prescribing `staging` and `production` environment labels is a big ask of Hokusai adopters, and could maybe be generalized to `--env production`-style arguments, but that's a separate discussion.)

And even if the local/remote command organization has drawbacks, it could still be helpful for Hokusai's internals (especially if we end up consolidating the ambiguous commands and basing their behavior on an env flag).

izakp commented:
Thanks all for the feedback.

@joeyAghion Re: `(I tend to agree that prescribing staging and production environment labels is a big ask of Hokusai adopters, and could maybe be generalized to --env production-style arguments, but that's a separate discussion.` I started following this pattern long-ago, and consider Hokusai as prescribing an "opinionated consistent workflow" - an organization can always alias "staging" and "production" kubectl contexts in their kubectl config file to whatever they like.

However, given that the feedback here seems to be against the `local` and `remote` grouping (I agree with @cavvia when you are skeptical of `our use of environments as both options and top-level command namespaces and the inconsistency arising from that`.

I am going to look into Click's [advanced patterns](http://click.pocoo.org/5/advanced/) and see if i can't get the environments loaded dynamically from the kubeconfig file.

izakp commented:
Okay all - I think you'll like this.  Dropped the `local` and `remote` groups and added `staging`, `production` and `pipeline` groups to the top-level namespace.

This drops the `--staging` and `--production` flags.

The `pipeline` group deals with commands that compare the `staging` and `production` environments explicitly.  I think it compares intuitively to Heroku's pipelines.

Given the fact that we are already opinionated about the `staging` / `production` contexts, and this is a requirement for `pipeline` commands to function properly, I think we can address extensibility later.

```
isacpetruzzi@Artsy-Isac lewitt-api (master) $ hokusai
Usage: hokusai [OPTIONS] COMMAND [ARGS]...

  Hokusai is a CLI for managing application deployments on Kubernetes

Options:
  --help  Show this message and exit.

Commands:
  build       Build the Docker image defined in...
  check       Check Hokusai dependencies and configuration
  configure   Install and configure kubectl
  dev         Interact with docker-compose development...
  pipeline    Interact with the project's' staging ->...
  production  Interact with the production Kubernetes...
  registry    Interact with the project registry defined by...
  setup       Set up Hokusai for the current project
  staging     Interact with the staging Kubernetes...
  test        Boot the docker-compose test environment...
  version     Print Hokusai's version and exit


isacpetruzzi@Artsy-Isac lewitt-api (master) $ hokusai dev
Usage: hokusai dev [OPTIONS] COMMAND [ARGS]...

  Interact with docker-compose development environment defined by
  ./hokusai/development.yml

Options:
  --help  Show this message and exit.

Commands:
  clean   Stop and remove all containers in the...
  logs    Print logs from the development environment...
  run     Run a command in the development...
  start   Start the development environment defined in...
  status  Print the status of the development...
  stop    Stop the development environment defined in...


isacpetruzzi@Artsy-Isac lewitt-api (master) $ hokusai registry
Usage: hokusai registry [OPTIONS] COMMAND [ARGS]...

  Interact with the project registry defined by `./hokusai/config.yml`

Options:
  --help  Show this message and exit.

Commands:
  images  Print images and tags in the project's remote...
  push    Build and push an image to the project's...


isacpetruzzi@Artsy-Isac lewitt-api (master) $ hokusai staging
Usage: hokusai staging [OPTIONS] COMMAND [ARGS]...

  Interact with the staging Kubernetes environment defined by the `staging`
  context in `~/.kube/config` and the  Kubernetes resources in
  `./hokusai/staging.yml`

Options:
  --help  Show this message and exit.

Commands:
  create   Create the Kubernetes remote resources...
  delete   Delete the Kubernetes remote resources...
  deploy   Update the project's deployment(s) to...
  env      Interact with the runtime environment for the...
  history  Print the project's deployment history in...
  logs     Get container logs
  refresh  Refresh the project's deployment(s) by...
  run      Launch a new container and run a command
  status   Print the Kubernetes remote resources status...
  update   Update the Kubernetes remote resources...


isacpetruzzi@Artsy-Isac lewitt-api (master) $ hokusai staging env
Usage: hokusai staging env [OPTIONS] COMMAND [ARGS]...

  Interact with the runtime environment for the application

Options:
  --help  Show this message and exit.

Commands:
  create  Create the Kubernetes configmap object...
  delete  Delete the Kubernetes configmap object...
  get     Print environment variables stored on the...
  set     Set environment variables - each of...
  unset   Unset environment variables - each of...


isacpetruzzi@Artsy-Isac lewitt-api (master) $ hokusai production
Usage: hokusai production [OPTIONS] COMMAND [ARGS]...

  Interact with the production Kubernetes environment defined by the
  `production` context in `~/.kube/config` and the  Kubernetes resources in
  `./hokusai/production.yml`

Options:
  --help  Show this message and exit.

Commands:
  create   Create the Kubernetes remote resources...
  delete   Delete the Kubernetes remote resources...
  deploy   Update the project's deployment(s) to...
  env      Interact with the runtime environment for the...
  history  Print the project's deployment history in...
  logs     Get container logs
  refresh  Refresh the project's deployment(s) by...
  run      Launch a new container and run a command
  status   Print the Kubernetes remote resources status...
  update   Update the Kubernetes remote resources...


isacpetruzzi@Artsy-Isac lewitt-api (master) $ hokusai production env
Usage: hokusai production env [OPTIONS] COMMAND [ARGS]...

  Interact with the runtime environment for the application

Options:
  --help  Show this message and exit.

Commands:
  create  Create the Kubernetes configmap object...
  delete  Delete the Kubernetes configmap object...
  get     Print environment variables stored on the...
  set     Set environment variables - each of...
  unset   Unset environment variables - each of...


isacpetruzzi@Artsy-Isac lewitt-api (master) $ hokusai pipeline
Usage: hokusai pipeline [OPTIONS] COMMAND [ARGS]...

  Interact with the project's' staging -> production pipeline

Options:
  --help  Show this message and exit.

Commands:
  gitdiff  Print a git diff between the tag currently...
  gitlog   Print a git log between the tag currently...
  promote  Update the project's deployment(s) on...
```

izakp commented:
@cavvia I am ready to merge if you all are happy with this

izakp commented:
These PRs failed as I didn't run `./distribute.sh` yet and push version 0.4.0.  Will rebuild them when this is merged and I do.

cavvia commented:
This structure looks good to me! I'm not familiar with the `pipeline` syntax but happy to go with that for commands that compare staging to production.

How does `production deploy` differ from `pipeline promote`? I'm not clear on whether the `deploy` subcommands always try to push a new image or not.

izakp commented:
@cavvia `production deploy` deploys a given image tag (must be on ECR) to production, just as it does on staging.  The only difference there is in the flags now as command groups, i.e. `hoksuai deploy --staging <IMAGE TAG>` vs `hokusai staging deploy <IMAGE TAG>` and `hokusai deploy --production <IMAGE TAG>` vs `hokusai production deploy <IMAGE TAG>`

izakp commented:
I'm gonna merge this as I want to unblock those downstream PRs.

cavvia commented:
:+1: understood
